packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,339 @@
+import {
+  UserAction,
+  UserActionsState,
+  UserActionTimeline,
+  UserActionType,
+  UserSession,
+} from './insight-user-actions-state';
+
+const SECOND = 1000;
+const MINUTE = SECOND * 60;
+const SESSION_INACTIVITY_THRESHOLD_IN_MS = 30 * MINUTE;
+const NUMBER_OF_PRECEDING_SESSIONS = 2;
+const NUMBER_OF_FOLLOWING_SESSIONS = 2;
+
+export interface RawUserAction {
+  name: string;
+  value: string;
+  time: string;
+}
+
+/**
+ * Preprocesses the raw user actions data from the API into a user actions timeline with the following steps:
+ * - The data is first sorted by most recent and mapped into user actions.
+ * - The actions are then split into sessions based on the ticket creation date.
+ * - The sessions are then stored in the timeline object either as preceding, following or the case creation session.
+ * - The timeline is then filtered to exclude custom actions passed in the state `excludedCustomActions` and empty search queries.
+ * Note: The filtering is done at the end to avoid filtering out actions that would lead to incorrectly split a session into two by creating a gap.
+ * @param state {UserActionsState} - The state of the user actions
+ * @param actions {rawUserAction[]} - The raw user actions array to preprocess
+ * @returns {UserActionTimeline}
+ */
+export const preprocessActionsData = (
+  state: UserActionsState,
+  actions: Array<RawUserAction>
+): UserActionTimeline => {
+  if (!state.ticketCreationDate || actions.length === 0) {
+    return {
+      precedingSessions: [],
+      session: undefined,
+      followingSessions: [],
+    };
+  }
+
+  const ticketCreationdate = new Date(state.ticketCreationDate);
+  const ticketCreationTimestamp = ticketCreationdate.getTime();
+  const excludedCustomActions = state.excludedCustomActions;
+
+  const mappedAndSortedActions = mapAndSortActionsByMostRecent(actions);
+
+  const timeline = splitActionsIntoTimelineSessions(
+    mappedAndSortedActions,
+    ticketCreationTimestamp
+  );
+
+  const filteredTimeline = filterTimelineActions(
```

Comment:
There is a problem with the filtering, if for example a session contains only actions the would be excluded, the all the session should be removed.

there is no point of having a session that contains 0 action.

Reply:
A solution could be to avoid even adding an excluded action to a session at the first place.

Reply:
fixed that by re-working the filtering logic and moving it before we insert a session in the timeline.

Reply:
Like I previously mentioned, `insertSessionInTimeline` needs to contain a check that checks if the session is empty or not,
if it's not empty, add it to the timeline,
if it is empty, do nothing.

 
that's the solution that would prevent us from having empty sessions to the timeline.

Also please add unit tests for that covers these issues I'm telling you about to avoid fixing this twice.
Make sure you have a test that would fail before implement the solution and that would pass after implementing it.

Reply:
added a check for empty sessions